python/python/bystro/prs/preprocess_for_prs.py

```diff
@@ -71,39 +71,40 @@ def _load_genetic_maps_from_feather(map_directory_path: str) -> dict[str, pd.Dat
     return genetic_maps
 
 
-def _extract_nomiss_dosage_loci(dosage_matrix_path: str) -> pd.DataFrame:
+def _extract_nomiss_dosage_loci(dosage_matrix_path: str, score_loci: set, chunk_size=1000) -> set:
```

Comment:
I think this is not quite the right way to go about this.

What you want is to use the filter operation, because 1) there are opportunities to skip reading data (called predicate pushdown; won't work for us because of how we've structured the locus column, but in the future we will change this), 2) we keep the filtering in arrow, which will be faster and more memory efficient than converting to pandas first

Reply:
```suggestion
def _extract_nomiss_dosage_loci(dosage_matrix_path: str, score_loci: set, chunk_size=1000) -> set:
    """
    Reads a dataset from the provided dosage matrix Feather file path and filters out
    rows where any column other than 'locus' contains a missing dosage value (represented by -1)
    and loci not in the base GWAS summary statistics dataset.

    Args:
    ----
    dosage_matrix_path (str): The path to the Feather file containing the target dataset.
    score_loci (set): A set of loci from a base GWAS summary statistics dataset.

    Returns:
    -------
    set: A set containing loci with no missing dosage values across all other columns.
    """
    dataset = ds.dataset(dosage_matrix_path, format="feather")
    filtered_loci = set()

    # Creating a filter for score_loci
    score_loci_filter = pc.field("locus").isin(pa.array(list(score_loci)))

    samples = [name for name in dataset.schema.names if name != "locus"]
    filters = [pc.field(sample) >= 0 for sample in samples]

    # Combine the filters using boolean and logic
    combined_filter = filters[0]
    for f in filters[1:]:
        combined_filter = pc.and_(combined_filter, f)

    for batch in dataset.filter(score_loci_filter).to_batches(batch_size=chunk_size):
        chunk = batch.filter(combined_filter).to_table()

        filtered_loci.update(chunk["locus"].to_pylist())

    return filtered_loci```

Reply:
The suggestion needs to be tested, but is more idiomatic, and efficient

Reply:
My version also happens to be much faster; leverages multithreading

Reply:
[test_memory_usage_extract.py.zip](https://github.com/user-attachments/files/15537585/test_memory_usage_extract.py.zip)

```
pip install memory-profiler
time  mprof run test_memory_usage_extract.py &> test_memory_usage_extract_ak_suggestion.log
```